File: src/main/java/typeSpecifications/transducerSpecification/TransducerGraph.java
package typeSpecifications.transducerSpecification;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

// class for specifying the transducer in form of a graph. it consists of TransducerNodes and TransducerEdges.
// so far we can add nodes or edges to it and print it.
public class TransducerGraph {

    public Set<TransducerNode> nodes;

    public TransducerGraph() {
        nodes = new HashSet<>();
    }

    public void addTransducerObjectNode(TransducerNode... n) {
        nodes.addAll(Arrays.asList(n));
    }

    public void addTransducerObjectEdge(TransducerNode source, TransducerNode target, String incoming, String outgoing, double cost) {
        nodes.add(source);
        nodes.add(target);

        // don't add duplicate edges!

        for (TransducerEdge edge : source.getEdges()) {
            if (edge.getSource() == source &&
                    edge.getTarget() == target &&
                    edge.getIncomingString().equalsIgnoreCase(incoming) &&
                    edge.getOutgoingString().equalsIgnoreCase(outgoing) &&
                    edge.getCost() == cost) {
                return;
            }
        }
        source.getEdges().add(new TransducerEdge(source, target, incoming, outgoing, cost));

    }

    public Set<TransducerNode> getNodes() {
        return nodes;
    }

    // TODO: print isolated nodes. (same as in the query)
    public void printGraph() {
        for (TransducerNode node : nodes) {
            for (TransducerEdge edge : node.getEdges()) {
                edge.print();
            }
        }
    }


    // only use this for test purposes!
    public boolean equals(TransducerGraph otherGraph) {
        if (this.nodes == null || otherGraph.nodes == null) {
            return false;
        }

        if (this.nodes.size() != otherGraph.nodes.size()) {
            return false;
        }

        return equalsOtherNodeSet(otherGraph.nodes);
    }

    private boolean equalsOtherNodeSet(Set<TransducerNode> otherNodeSet) {
        int pairsFound = 0;
        for (TransducerNode thisNode : this.nodes) {
            for (TransducerNode otherNode : otherNodeSet) {
                if (thisNode.equals(otherNode)) {
                    pairsFound++;
                    break;
                }
            }
        }
        return pairsFound == (this.nodes.size());
    }
}

File: tin v2/src/main/kotlin/tin/model/transducer/TransducerGraph.kt
package tin.model.transducer

import java.util.*

class TransducerGraph {
    var nodes: MutableSet<TransducerNode> = HashSet()

    fun addTransducerObjectNode(vararg n: TransducerNode) {
        nodes.addAll(listOf(*n))
    }

    fun addTransducerObjectEdge(source: TransducerNode, target: TransducerNode, incoming: String, outgoing: String, cost: Double) {
        nodes.add(source)
        nodes.add(target)

        // don't add duplicate edges!
        for (edge in source.edges!!) {
            if ((edge.source.equals(source)) && (edge.target.equals(target)) &&
                    edge.incomingString.equals(incoming, ignoreCase = true) &&
                    edge.outgoingString.equals(outgoing, ignoreCase = true) && (edge.cost == cost)) {
                return
            }
        }
        source.edges.add(TransducerEdge(source, target, incoming, outgoing, cost))
    }

    fun getNodes(): Set<TransducerNode> {
        return nodes
    }

    // TODO: print isolated nodes. (same as in the query)
    fun printGraph() {
        for (node in nodes) {
            for (edge in node.edges!!) {
                edge.print()
            }
        }
    }

    // only use this for test purposes!
    fun equals(otherGraph: TransducerGraph): Boolean {
        return if (nodes.size != otherGraph.nodes.size) {
            false
        } else equalsOtherNodeSet(otherGraph.nodes)
    }

    private fun equalsOtherNodeSet(otherNodeSet: Set<TransducerNode>?): Boolean {
        var pairsFound = 0
        for (thisNode in nodes) {
            for (otherNode in otherNodeSet!!) {
                if (thisNode.equals(otherNode)) {
                    pairsFound++
                    break
                }
            }
        }
        return pairsFound == nodes.size
    }
}
